working/control-flow-collections/feature-specification.md

```diff
@@ -0,0 +1,976 @@
+    }
+    ```
+
+    This may fall down if the two closures are complex but I think that's rare
+    in practice.
+
+I think we probably *don't* want to blanket apply all of these fixes without
+user intervention. There may be some style preferences or the fix may not
+always succeed. This is a big enough change where having a human validate it is
+a good idea.
+
+## Next Steps
+
+As always, the immediate next step of a proposal is running it past the language
+leads and stakeholders.
+
+### Usability studies
+
+This feature has some good things going for it:
+
+*   It is mostly syntax sugar. A front end should be able to compile this down
+    to existing Dart semantics (with perhaps some extra support needed for `if`
+    in const collections). It shouldn't significantly impact the runtime or
+    backends, so the implementation cost should be relatively low.
+
+*   The semantics are narrow and fairly straightforward. It doesn't interact
+    with the type system in complex ways. It doesn't touch tricky parts of the
+    grammar, calling conventions, or runtime behavior. I think the
+    implementation is fairly low-risk. I don't think we're likely to run into
+    major surprises we didn't anticipate during implementation.
+
+However, the human side of this is less certain. I've tried to make the behavior
+intuitive by piggy-backing on syntax users already understand. But I worry that:
+
+*   Users will find it confusing to see `if` or `for` inside a collection
+    literal.
+
+*   Even after understanding it, users may not *like* the syntax.
+
+*   Users may be unhappy that the syntax doesn't go *far enough*. This feature
+    may lead them to expect, say, `while` to work inside a collection. They may
+    expect to be able to put an entire block of statements as the body of an
+    `if`. They may want to use `if` outside of a collection but in other
+    expression contexts.
+
+    In other words, this may be a "garden path" feature that encourages a whole
+    set of expectations, some of which are met and the rest of which are
+    confounded.
+
+*   Users may want to include multiple elements inside a body and not know how
+    to accomplish that. The spread proposal gives them a mechanism, but it may
+    not be a natural or obvious one.
+
+I don't think we can reasonably resolve these on paper, so before committing to
+this proposal, I think we should do user studies of some of these scenarios and
+then tune this proposal based on how those go.
+
+### Conditional arguments
+
+This proposal only covers conditional execution in collections. A natural
+extension that would be particularly useful for Flutter is to extend it to
+argument lists:
+
+```dart
+IconButton(
+  icon: Icon(Icons.menu),
+  tooltip: 'Navigation menu',
+  if (isAndroid) padding: const EdgeInsets.all(20.0),
+)
+```
+
+Without [rest parameters][], `for` isn't useful and `if` probably isn't
+feasible for positional arguments. But even without rest params, it's possible
```

Comment:
What if positional arguments are made optionally "nameable"? Could that help with feasibility?

Reply:
...yes, but making positional arguments optionally nameable also has its own feasibility problems. :) That's a fairly gnarly design problem, I think.